databuilder/tables/beta/tpp.py

```diff
@@ -115,6 +131,28 @@ class addresses(EventFrame):
     care_home_requires_nursing = Series(bool)
     care_home_does_not_require_nursing = Series(bool)
 
+    def active_for_patient_at(self, date):
+        """
+        Return each patient's "active" address as of the supplied date
+        """
+        spanning_addrs = self.take(self.start_date <= date).drop(self.end_date <= date)
+        # Where there are multiple overlapping address registrations we need to pick one.
+        # Logic copied from:
+        # https://github.com/opensafely-core/cohort-extractor/blob/e77a0aa2/cohortextractor/tpp_backend.py#L1756-L1773
+        ordered_addrs = spanning_addrs.sort_by(
+            # Prefer the most recently registered address
+            self.start_date,
+            # Prefer the address registered for longest
+            self.end_date,
+            # Prefer addresses with a postcode
+            case(when(self.has_postcode).then(1), default=0),
+            # Use the opaque ID as a tie-breaker for sort stability (we invert this simply
+            # so the order matches the original order defined in Cohort Extractor to
+            # facilitate direct comparison)
+            -self.address_id,
```

Comment:
This logic was ported from Cohort Extractor but, looking at it now, I'm not convinced it's what we want. I think we might want to use `has_postcode` as the _most_ important sort condition (because I think that's what fields like `MSOA` or the care home fields hang off) and only after that use the dates.

Reply:
@inglesp and I decided that preferring addresses with postcodes was the right call because everything useful associated with an address (MSOA, IMD, care home status etc.) requires having a postcode. And now we're no longer trying to maintain exact parity with Cohort Extractor, we can drop the silly negation on `address_id`.